### ISSUE TITLE
wait until Vsync API

### ISSUE THREAD

**Issue by ad8e** [glfw/glfw#1157]
I'd like to do my rendering and processing just before each Vsync, to reduce latency. So if a frame takes 0.1s, then my application will wait for 0.08s after the last Vsync, then render and process everything in the next 0.01s, so that the frame is only 0.02s old when it is pulled. This is only possible if GLFW has a function for reporting when the Vsync happened, such as GetTimeSinceLastVSync or WaitForVSync. With this information, I can measure the time between frames, measure my render loop's time, and then insert my rendering at the appropriate location just before the Vsync.

The advantage over single buffering is less CPU usage, and the advantage over double buffering is less latency.

VR implementation: https://developer.oculus.com/blog/optimizing-vr-graphics-with-late-latching/ calls it "late-latching" and states that consoles have had it for a long time
VR implementation: http://media.steampowered.com/apps/valve/2015/Alex_Vlachos_Advanced_VR_Rendering_GDC2015.pdf P20-22. uses [IDXGIOutput::WaitForVBlank](https://msdn.microsoft.com/en-us/library/windows/desktop/bb174559(v=vs.85).aspx?f=255&MSPPError=-2147217396#) on Windows. Their presentation talks about GPU rendering start points which is different, and they want the end of VSync rather than the beginning, but their timepoint API serves the same purpose, lowering latency.
recent: https://gamedev.stackexchange.com/q/151893

**Comment by ferreiradaselva** [comment]
Does any of the underlying API (EGL, GLX, WGL) have a function for getting that time? Those examples (in the links) look like just a second thread for rendering, using a [fence](https://www.khronos.org/opengl/wiki/Sync_Object#Fence) and calculating the start time and end time.

If there's no such GL function that makes this simpler (not using fence), I don't see why including this in GLFW. It would only be useful on multi-threaded applications, would affect single-threaded applications, and would require creating OpenGL objects that the user don't know about.

**Comment by ad8e** [comment]
I don't know how this is done at all. The reason it seems logical for this to be a GLFW thing is that GLFW handles all my OpenGL-related framebuffer stuff, and this feature supersedes GLFW's other framebuffer functions. If it is really as you say, that implementing this requires a second thread, then perhaps it doesn't belong in GLFW.

**Comment by ferreiradaselva** [comment]
> If it is really as you say, that implementing this requires a second thread, then perhaps it doesn't belong in GLFW.

It doesn't necessarily need to be on a second thread, but for the situation you want, it makes sense only if the rendering is done on a second thread. On a single thread, you start the swap operation on a specific time (start/end of the main loop), without blocking your application. Using a fence ([`glFinish`](https://www.khronos.org/registry/OpenGL-Refpages/gl4/html/glFinish.xhtml) for `glSwapBuffers`) would mean to block until the GPU is done, then take the time difference. On a second thread, that operation wouldn't be a problem, but on a single-threaded application, it means that the GPU will block the main thread until it's done, which could be a hit on performance.

**Comment by tombsar** [comment]
I've had reasonable results in the past by simply recording the time (e.g. `last_swap_time = glfwGetTime()`) immediately after any call to `glfwSwapBuffers()`, and predicting the next swap time as that plus the frame interval. I don't know of any API to get a more accurate number than that, unfortunately.

**Comment by ad8e** [comment]
@ferreiradaselva, Yes, your concern is valid. But this can be good even without threading, because the extra 0.08s gained captures input. For applications which exceed the monitor framerate, the lost sync time between GPU->framepoint is fine because waiting idle for input is the ideal situation anyway. For example, consider an application like tetris; it doesn't need to do anything computationally demanding, but "late-latching" produces more responsive input. Its best option between the end of rendering and 0.08s into the next frame, is to wait for input.

If your application requires the full frame to process or it will run out of time -> need two threads to use "late-latching", or trouble starts
If your application requires less than a full frame to process -> no need for two threads, late-latching produces better results automatically

**Comment by ad8e** [comment]
I've deleted my previous reply (which came after tombsar's comment) because now I have a better understanding of where things are, after testing.

@tombsar says the timepoint can be acquired by waiting after glSwapBuffers, but that's not quite right. Vsync needs to be enabled for this to be true. With glfwSwapInterval(0), I get all my frames one after another. glfwSwapInterval(1) produces the right timepoints, but that defeats the whole point, since now every frame is delayed by one.

The right thing to compare the proposed API to is double buffering with Vsync. glfw uses glfwSwapInterval(1) to implement Vsync, which delays each frame by one. With double buffering and Vsync, glfw already has the block-until-Vsync behavior as it waits for the buffer to clear up. After blocking until the next Vsync, glfw waits until the second Vsync to show the image, producing 0.2s of delay between input and display, using our example. Since with Vsync we can use @tombsar's method to approximate timepoints, this can be cut down to 0.12s by delaying rendering, and that's the best possible. With the proposed API, glfw has the same blocking as double buffer vsync, except that now the delay is 0.1s automatically, and can be cut down to 0.02s. Single threaded mode doesn't impose any cost since the delay is equal in both cases.

The only disadvantage of the proposed API over double buffering w/ Vsync is that if rendering stutters and overlaps with the Vsync point, then there will be tearing as opposed to frame drops. This can be worked around with double buffers+timing if the user finds it an issue, producing strictly better performance, but with increased complexity.

Conceptually, glfwSwapInterval(1) waits for Vsyncs to know when to swap, but the proposed API gives better swap deadlines to reduce latency.

**Comment by ghuser404** [comment]
I'd think why the consoles have had it for a long time is because a particular console comes with specific hardware, and as a developer you can rely on your frames being processed within the same time for all your users. This is not true for desktop/mobile platforms, however.

While I'm not saying this feature is useless, did you consider using fences, as @ferreiradaselva suggested?

P.S. Do macOS/Linux support similar functionality? Otherwise, this could be considered for platform-specific API.

**Comment by ad8e** [comment]
I'm not familiar with consoles. But the uniformity of the hardware doesn't fix things automatically; you still need an API to find the Vsync point. Even if all your renders took exactly the same time as a single monitor frame, it wouldn't help because they'd just land in a consistent random spot within a frame (possibly having tearing on every frame).

Fences don't appear to work here. (Otherwise, why would the VR people go through such hoops?) The timepoints don't exist unless glfwSwapInterval(1) is turned on, and then everything is delayed.

MacOS has [CVDisplayLinkOutputCallback](https://developer.apple.com/documentation/corevideo/cvdisplaylinkoutputcallback), but Apple doesn't document it so I don't know if its behavior is appropriate.

The docs on Linux I could find was [this presentation on KMS/DRM, page 25](https://events.linuxfoundation.org/sites/events/files/slides/brezillon-drm-kms.pdf). I'm unsure if that's available to users.

**Comment by ghuser404** [comment]
@ad8e, if you knew how long the sync takes exactly, you could do the swap with interval 1, then change it to 0 interval after the swap. After that, you would know when every frame will get swapped.

**Comment by ad8e** [comment]
I'm planning on that (ignoring drift and monitor changes), but glfw's docs say "swap interval extensions used by GLFW do not allow the swap interval to be reset to zero once it has been set to a non-zero value." So this solution has drawbacks.

**Comment by tombsar** [comment]
FWIW I would love an API to do this, there just doesn't seem to be any feasible solution.

Perhaps https://www.x.org/releases/X11R7.7/doc/libXext/synclib.txt gives a clue for how to get this information on X11? I'll try to play with it, but don't hold much hope...

**Comment by tomc1998** [comment]
Regarding linux:

I can confirm `GLX_SGI_video_sync` works, kde also uses `GLX_OML_sync_control` which together cover most linux configs. There's also the `Present` [extension for x11](https://cgit.freedesktop.org/xorg/proto/presentproto/tree/presentproto.txt), possibly glx uses this internally.

[kde SGI usage](https://invent.kde.org/plasma/kwin/-/blob/master/src/backends/x11/standalone/x11_standalone_sgivideosyncvsyncmonitor.h)

[kde OML usage](https://invent.kde.org/plasma/kwin/-/blob/master/src/backends/x11/standalone/x11_standalone_omlsynccontrolvsyncmonitor.h)

Wayland has the [presentation time protocol](https://wayland.app/protocols/presentation-time). Looks like this was implemented in glfw in #1406.

Would certainly be nice to have support for this, especially because it's not possible to hook into raw x11 events from outside glfw, meaning I can't use x11's `present` extension without patching glfw.

**Comment by ad8e** [comment]
Oh neat!

For Windows: Windows has the function D3DKMTGetScanLine. Instead of waiting for VBlank, it simply queries "what scanline is the monitor on" and then returns the number immediately. In my testing, this is significantly more accurate than D3DKMTWaitForVerticalBlankEvent. It's also much easier to use, with a simple linear regression, implementable in 2 hours, outperforming my complex dynamic programming solution that took weeks to theorize. It's also much faster on my system. So it's benefits all around.

However, Mark Rejhon (ChiefBlurBuster) tested D3DKMTGetScanLine on his 1080 Ti and reported two things: it takes 8 scanlines to run (which is very slow), and "the scanline counter doesn't increment properly in the porch (when InVerticalBlank is true)". His guess is that D3DKMTGetScanLine does not play well with Nvidia cards, though thorough testing has not been done.

I have all this code available, though it's reliant on other libraries I have, so let me know if anyone wants it.

For Linux: I don't know of any comparable function to D3DKMTGetScanLine. Maybe this is an issue for the Wayland folks to investigate.

For GLFW: I think a "wait until Vsync" API is not appropriate, since it's not the best choice on Windows. A user would want to switch between Wait and GetScanLine for each platform. And GetScanLine requires acquiring the porch information over time, so it's not a simple API.

**Comment by ad8e** [comment]
What I said about Linux not having a comparable to D3DKMTGetScanLine is wrong. I looked at OML_sync_control and it appears to correspond to GetScanLine. I might test this on my graphics card, but it won't happen soon (months to a year).

**Comment by tomc1998** [comment]
You don't need the scan line for low latency vsyncing like OP was talking about, you just need to know when the vblank starts (and ends?) - if the monitor is currently scanning, then you don't want to render

OML_sync_control isn't available on everything, my rtx 2070 only supports the SGI extension, but my intel 5000 HD laptop supports both OML & SGI

**Comment by ad8e** [comment]
Right, the scanline is not necessary. It's just that on Windows, the only wait-free way to get the Vblank start point also gives the scanline. So it's necessary to do a small amount of scanline math to recover the VBlank start point.

I suspect the OML version will be preferred when available, but I have a pretty good fallback algorithm for Wait too.

**Comment by tomc1998** [comment]
I think waiting is fine, what's your use case for a non-waiting solution?

I *believe* kwin chooses SGI over OML, it did a year ago at least, unsure why. Perhaps getting the scanline itself adds significant overhead - worth testing.

**Comment by ad8e** [comment]
Benefits of non-waiting over waiting:

Accuracy. It's way more precise, on my Intel. Waiting solutions have late wakeups which can delay the timepoint by large amounts.

Ease of use. Extracting the period and phase of the VBlank signal is very hard, when the wakeup timepoint can be delayed. Here's my solution:
[vsync.txt](https://github.com/glfw/glfw/files/10710886/vsync.txt). It's many pages just to describe the algorithm, and it isn't even finished, needing a way to recover gracefully from delays lasting more than a frame, which are common in practice. This took about 5 weeks to write.

Performance. It's almost instant, on my Intel graphics card.

No need for separate thread. It can run directly on the render thread since it's instant.

Kwin's OML uses its Wait version. Probably just from lack of time; I know its new presentation code, while much better than its old version, still has some inefficiencies and poor heuristics. The OML code doesn't use the scanline; it gives the timepoint of the VSync. See https://registry.khronos.org/OpenGL/extensions/OML/GLX_OML_sync_control.txt, glXGetSyncValuesOML(). UST is the timestamp, MSC and SBC count the # of retraces and swaps respectively.

**Comment by mdrejhon** [comment]
Hola,

Author of TestUFO here (www.testufo.com) which was extensively used by the kwin-lowlatency fork to get VSYNC working properly in web browsers on 144Hz+ monitors.

Sometimes you don't want waiting and sometimes you want waiting -- both options should be provided because sometimes precision wait-control reduces latency by a giant amount.  ad8e hits the nail on the heads.

There are some interesting lag-increasing and lag-reducing nuances (e.g. Quick Frame Transport, Variable Refresh Rate, etc) where you want a blocking wait until near beginning of a refresh cycle, instead of at end of refresh cycle.

So, my purpose of my post:

# "Why does precise scanline control reduce VSYNC ON input lag by up to a full refresh cycle?"

## Excellent Use Case: Whenever Linux is used for VR

## This is what the Android implementations already do, so you open source people need to read carefully and understand.  Low lag is super important to avoid dizziness and vertigo.   

Some of us may whine silly about whether milliseconds matter or not, but a good read-up is [The Amazing Human Visible Feats of The Millisecond](https://forums.blurbusters.com/viewtopic.php?t=6808) that I made in my forum;

Lag is very important in some contexts (e.g. emulating a real life Holodeck)

For example, most 240Hz monitors can do a custom Linux ModeLine that creates a gigantic VBI for a lower fixed-Hz, **e.g. 60Hz refresh cycles transmitted over the video cable in 1/240sec.** ... basically the same Pixel Clock and Horizontal Scanrate.

Normally named by HDMI Forum:

![image](https://user-images.githubusercontent.com/59981975/218194947-4c42f1a7-3ab3-4841-bb09-032650289896.png)

But we found out that in an undocumented way, this works fine over DVI and DisplayPort and USB-C Altmodes, on some displays, by creating custom Linux ModeLines (or for, Windows ToastyX CRU EDID overrides into the Windows Registry)

So by forcing a 60Hz to a higher horizontal scanrate by using ultra-large blanking intervals.  (The boom of VRR means most displays now have undocumented QFT support, since VRR is defacto QFT, and QFT can be metaphorically viewed as simply a fixed-Hz version of VRR)

Most diplays behave like this:

![image](https://user-images.githubusercontent.com/59981975/218196927-1bacd88f-9f5d-4705-97ce-2e3402c57620.png)

QFT speeds up delivery of frames:

![image](https://user-images.githubusercontent.com/59981975/218196951-5c63a949-a6b2-453a-8190-39f3fbe640fd.png)

(Ignore the 2000fps part as that referred to the light/dark grey bars representing frameslices, these images were used in a different context.  These images are still excellent illustrations of QFT)

Some displays scanout in sync with the cable (with a tiny rolling display-processing buffer), and some displays buffer the full refresh cycle before scanning out.   Both displays benefit from QFT in subtly different ways, but understanding why QFT reduces lag in multiple different ways, is key.

Some 60Hz displays can accept a frame transmission over the HDMI or DisplayPort cable in 1/120sec, but it's fairly undocumented.  However, more and more QFT EDIDs are becoming 

Some specialized displays (such as VR headsets) already do this internally, but many end users at Blur Busters have done custom EDIDs and custom ModeLines make this possible on several panels.

![image](https://user-images.githubusercontent.com/59981975/218196638-cc67ab18-3156-4dd7-a643-0243dc376ca0.png)

![image](https://user-images.githubusercontent.com/59981975/218200736-804d68e9-d7e7-497e-ad14-a35a5e145900.png)

So sometime scanline-waiting control is quite handy here, to decide how early/late in the horizontal scanrate to release a frame to, which upon a precisely delayed-as-possible wait-unblocking, can cause the software to do an inputread closer to the next refresh cycle, and thus reduce input lag (input-to-photons).

There is a LOT that software can do to reduce input lag, including precise scanline-waiting control (which is why Windows-based RTSS Scanline Sync was invented), but that is often more advanced programming for those who don't understand this bleep...

VSYNC OFF Tearlines are just rasters ([proof1](https://www.pouet.net/topic.php?which=11422&page=1). [proof2](https://forums.blurbusters.com/viewtopic.php?f=22&t=4213)).  So utilizing this, we have extolled the use of VSYNC OFF to simulate VSYNC ON (at the lowest possible input latency) by putting the "tearing" hidden right above the top edge of the screen.  Some PC software now uses this, based on our research.

![image](https://user-images.githubusercontent.com/59981975/218199947-e8ec8044-4f92-40bc-8365-4771a8853a2b.png)

(...Understanding how, during VSYNC OFF, that frame presentation "splices" the video signal at the current raster scanout, is quite handy background knowledge...)

This is simply background information on how some advanced "latency-priority" software developers (e.g. VR software programmers and esports game programmers) does.  Blur Busters specializes in the Present()-to-Photons chain, but from the OpenGL perspective, that's the glxxSwapBuffers()-to-photons chain.  As I work with display manufacturers, I understand that black box, so ask away your questions -- few people understand that black box.

BTW, from a future proofing POV, remember that sub-millisecond relative sync between input-time and photon-time is very important.  Whether to reduce stutter (2ms jitter error in 8000 pixels/sec translates to a 16 pixel stutter, which can become more visible at 8K 240Hz laboratory display than at 1024x768 60Hz).   (...Since VR requires a giant amount of pixels and FOV to look remotely realistic, it's a useful thing to keep in mind too...)

Now, I should add some cred:

To understand more about Blur Busters research, check out www.blurbusters.com/area51 where I'm cited in over 25 peer reviewed papers.  I'm here to give free advice on advanced lag-reducing software-based sync technologies.

Blur Busters also happens to be the website that extolls the 1000fps 1000Hz future; which is coming this decade (1000Hz prototypes are already in the lab).  A common misconception is that humans can't see past a specific Hz, when the nuances are more complex, e.g. 240fps on a flickerless 240Hz 0ms-GtG display is the same display motion blur as a 1/240sec SLR camera shutter photograph, and 1000fps 1000hz on a flickerless 1000Hz 0ms-GtG display is the same display motion blur as a 1/1000sec SLR camera photograph.  Flicker reduced motion blur on legacy displays (e.g. CRT).

As extra Hz becomes cheap (120Hz is a freebie in some displays now as of 2023, commoditization will happen to 240Hz too), I believe more open source initiatives should learn about the semi-mysterious black box between graphics driver frame releases and the photons hitting eye balls, and Blur Busters is one of the better teaching tools to learn about this specific black box.  Knowing this can improve motion quality (fewer stutters) and reduce input lag (important for some contexts, especially Linux VR).

(That's why refresh rate upgrades need to be ultra-geometric when it comes to flickerless displays such as LCD, OLED, etc.  I often recommend people upgrade refresh rates by 2x-4x to see noticeable differences)

**Comment by mdrejhon** [comment]
> I think waiting is fine, what's your use case for a non-waiting solution?
> 
> I _believe_ kwin chooses SGI over OML, it did a year ago at least, unsure why. Perhaps getting the scanline itself adds significant overhead - worth testing.

You don't need to get the scanline itself directly from the graphics driver.

You just load the current known horizontal scanrate from the current Linux ModeLine used on the display.  And use RTDSC instruction to monitor a time-offset between exact VBLANK-start or VBLANK-end timestamps.   From that, you can compute a very accurate approximate scanline number.  

That's what I used for Tearline Jedi (precise tens-of-microseconds raster-control of VSYNC OFF tearlines -- in Windows C#, even!).

All graphics APIs on all platforms, IMHO, should provide a raster poll (at the minimum), even if 'hacked' from this.

**Comment by mdrejhon** [comment]
> You don't need the scan line for low latency vsyncing like OP was talking about, you just need to know when the vblank starts (and ends?) - if the monitor is currently scanning, then you don't want to render

If the VBI is gigantic (3x the size of visible image), the location of release of frame can have a dramatic impact on latency (3/4ths of a refresh cycle latency difference between start-of-VBI release and end-of-VBI release).   

Since VSYNC ON graphics drivers don't let you control release timing, we often use VSYNC OFF as a workaround and beamsynced (raster scan line beam raced) release of frame.  See above.

**Comment by mdrejhon** [comment]
> However, Mark Rejhon (ChiefBlurBuster) tested D3DKMTGetScanLine on his 1080 Ti and reported two things: it takes 8 scanlines to run (which is very slow)

The penalty is partially caused by the fact that calling GetScanLine essentially causes some kind of partial graphics-flush event on the GPU, in order to get the raster.

It varies a lot between systems; sometimes fewer scanlines and sometimes more scanlines.  

Instead, a fast time-estimate can be done as time intervals between VSYNC's.  
On my RTX 3080 it performs differently.  

This is very easy to do with knowing VSYNC heartbeats (tens-microsecond-accurate timings of starts of refresh cycles, or timings of ends of refresh cycles) and a microsecond counter such as:

* C++ std::chrono::high_resolution_clock
* RTDSC instruction (all platforms)
* QueryPerformanceCounter() (Windows)
* micros() on Arduino/microcontrollers
* clock_gettime in POSIX (high resolution modes)
* etc.

Historically, Linux has had a problem knowing the VSYNC heartbeat of the graphics card, which is why things often stuttered on non-60Hz monitors in X in the past, until fixed in newer implementations such as kwin-lowlatency fork (and its various backports to main kwin).

Since horizontal scanrates is usually less than a million scanlines (pixel rows) per second, even tens of microsecond accuracy is okay on most platforms for raster estimation use-cases.

> and "the scanline counter doesn't increment properly in the porch (when InVerticalBlank is true)". His guess is that D3DKMTGetScanLine does not play well with Nvidia cards, though thorough testing has not been done.

I noticed that on some graphics cards, the scanline counter does increment in the blanking interval, but we can safely say that we cannot assume so.   In the abscence of this data, we estimate a raster as a time offset between refresh cycles (computed from the display's horizontal scanrate).

> ~For Linux: I don't know of any comparable function to D3DKMTGetScanLine. Maybe this is an issue for the Wayland folks to investigate.~

There's none available, but there are now APIs to check for blanking interval timing (VSYNC), which has solved a lot of problems for the kwin-lowlatency team, which has found ways to sync (even with those proprietary graphics drivers).  As long as you can pull a heartbeat from that, the software developer can estimate time offsets between blanking intervals for various use cases, most of which involves latency reductions.

> For GLFW: I think a "wait until Vsync" API is not appropriate, since it's not the best choice on Windows. A user would want to switch between Wait and GetScanLine for each platform. And GetScanLine requires acquiring the porch information over time, so it's not a simple API.

A compromise would be a poll check such as .InVBlank that says true if inside blanking interval (VBI), and false if outside.  

There are use cases where we need access to this independently of whether we're using VSYNC ON (waiting) or VSYNC OFF (not waiting).  Then raster scanline estimation can be left to the software developer.

**Comment by ad8e** [comment]
I got OML working on my Intel GPU, thanks to GLFW exposing native APIs. That means Linux and Windows support vsync timing with GLFW. 

The vsync APIs are so inconsistent across platforms that it's better to leave them out of GLFW. None of the setup functions are the same, none of the APIs return the same info, and they all require different management.

Here's some baseline OML code:

```
#define GLFW_EXPOSE_NATIVE_X11
#include "GLFW/glfw3.h"
#include "GLFW/glfw3native.h"
#define GLX_GLXEXT_PROTOTYPES //for glXGetSyncValuesOML
#include "GL/glx.h"

GLXDrawable global_drawable;
Display* global_display;
int64_t ust_global; //timestamp
int64_t msc_global; //vertical retrace number
int64_t sbc_global; //swap buffer number

//run this after making the OpenGL context current
void prepare_sync() {
	global_drawable = glXGetCurrentDrawable();
	global_display = glfwGetX11Display();
	outc(glXQueryExtensionsString(global_display, 0)); //0 for now, no idea how to get the screen
	outc("display drawable", global_display, global_drawable);
	//check(glfwExtensionSupported("GLX_OML_sync_control"), "OML not supported"); //this is not the right way to check for the extension
	//just assume it exists for now. todo, see https://invent.kde.org/plasma/kwin/-/blob/master/src/backends/x11/standalone/x11_standalone_omlsynccontrolvsyncmonitor.cpp
}

void get_sync_values() {
	bool result = glXGetSyncValuesOML(global_display, global_drawable, &ust_global, &msc_global, &sbc_global);
	check(result == 1, "OML failed");
	outc("UST was", ust_global, msc_global, sbc_global, now());
}
```

On my system, the timestamp is in microseconds. It appears aligned to the system time (clock_gettime). I hope it's the monotonic clock, and I hope it's exactly aligned, but I haven't checked.

**Comment by mdrejhon** [comment]
> The vsync APIs are so inconsistent across platforms that it's better to leave them out of GLFW. None of the setup functions are the same, none of the APIs return the same info, and they all require different management.

Fortunately, this is testable in multiple ways with a visual "tearing position" test.

To sync it up somewhat.  It's a workload, but would make possible cross-platform beam racing, once it was tested on the most common graphics drivers on the most common platforms.  

AD8E, you're already a honorary Tearline Jedi (successful precision VSYNC OFF tearline position control).   are you willing to do this work to analyze?   **(Note: Please email mark [at] blurbusters.com if you prefer a paid commission to turn this into paid research -- I fully offer this funded research to Apache, MIT or GIT)**   

This is something some developers in my community badly needs, to the point where I'm willing to fund it a bit, or BountySource it if someone has a "preferred bounty system"...

Crossplatform beam racing is a latency Holy Grail, and is definitely possible if we have access to VSYNC heartbeats during VSYNC OFF mode.  All platforms now provide that, but are very platform-specific.  Rolling this into a generic raster poll or a generic in-blanking poll, is a very big challenge that requires testing behaviors of existing implementations.

While it's possible GLFW may or may not be right venue, such capability still needs to be integrated into *some project*; so it might as well be a helper API of an existing project or a plugin/module to an existing graphics API wrapper.

# How To Test For VSYNC Timing Behaviors Between Different Platforms

1. Have a reasonably fast-enough GPU (AMD or NVIDIA) in your computer, even an old NVIDIA GTX 980 will be adequate at lower resolutions for testings' sake.  This will reduce error margins in this analytic testing, e.g. fast memory bandwidth and more instant graphics flushes

2. For this testing phase, prevent GPU power management from occuring by using performance mode

3. First, verify for proper VSYNC ON support (framerate=Hz capability).  Easiest is to test 50Hz or 75Hz with a simple fast horizontally scrolling animation.   
... Zero stutters = proper VSYNC ON support.  
... Stutters = API is using an internal clock 
Most desktop LCDs support either 50Hz and/or 75Hz, so that's a good non-60Hz reference to test for perfection of driver-level VSYNC ON support.  You will witness zero-stutters on PC platforms, but lots of stutters on some Linux platforms.  

4. [optional] If you're getting stutters, you need to manually find a way to sync by accessing the VSYNC heartbeat via an alternate API and rolling your own software-based sync rather than a driver-based sync.  APIs such as OML_sync_control (Linux) could help de-stutter an OpenGL implementation that doesn't have good non-60Hz sync.

5. Once you've solved VSYNC ON fluidity and you can do simple animations at perfect framerate=Hz sync (this is something achievable on all platforms nowadays, thankfully -- although a bit more fiddling than usual when it comes to non-Android Linux), move onto the implementation of a VSYNC-heartbeat-providing API to expose VSYNC timing to the user

6. Use VSYNC OFF and uncap the frame rate.  (glfwSwapInterval(0) in GLFW).  Verify that tearing appears, which is mandatory for some latency-reducing tricks.  The developer must be given the choice to have intentional tearlines, and then control/invisibilize the tearlines.  But don't worry about tearling-position control; **just give the option of making tearing appear with GLFW on all platforms** -- that is key.  Usually the desktop window manager gets in the way (makes tearing disappear), but usually full screen mode can make the tearing appear.  
*As long as this happens, then we leave it to the end user / developer to get access to the true-tearing mode for the low latency beam raced goodness as well as software-based sync technologies (tearingless VSYNC OFF similar to RTSS Scanline Sync app for Windows).*

7. Wrapper the existing APIs into a unified "Are We In Blanking Or Not?" API

* Linux: OML_sync_control(?) or other similar equivalent
* Windows: [D3DKMTGetScanLine](https://learn.microsoft.com/en-us/windows-hardware/drivers/ddi/d3dkmthk/nf-d3dkmthk-d3dkmtgetscanline)
* Mac: [CVDisplayLinkOutputCallback](https://developer.apple.com/documentation/corevideo/cvdisplaylinkoutputcallback)
* Android: [Choreographer.VsyncCallback](https://developer.android.com/reference/android/view/Choreographer.VsyncCallback)

Also, workarounds, if required:

* On some platforms, there may be no non-blocking method, so for those, you may need to run a separate thread for monitoring the VSYNC heartbeat.  

* Also there's the generic method: Relying on graphics drivers' timing (if graphics drivers support VSYNC): Platforms that support a simultaneous VSYNC ON buffer (swapinterval 1) and a VSYNC OFF buffer (swapinterval 0) in two offscreen buffers, use a dummy loop on an offscreen VSYNC ON buffer to pull the VSYNC heartbeats.   This apparently worked on some older versions of MonoGame C# code, although not on all platforms.

8. Run software in VSYNC OFF.  Run that panning test in VSYNC OFF mode, but dummy-wait-loop on the step 7 and then intentionally add a random refresh cycle delay (e.g. quarter refresh cycle delay, half refresh cycle delay), to randomly reposition an intentional stationary tearline.  Then present rendered frame (of panning bitmap) and immediately flush at the spot.

9. If you see a nearly-perfectly stationary VSYNC OFF tearline, we've now confirmed that platform is crossplatform beamraceable.  This was already confirmed on PC (DX, Vulkan, OpenGL) and Mac (Vulkan, OpenGL) as long as you got access to the tearline-visible mode.

____

Optional refinement (but recommended)

10. For determining how the VSYNC heartbeat is occuring relative to scanout for that specific platform or specific graphics driver -- we need to know if it's a start-of-VBI heartbeat or end-of-VBI heartbeat.   Most drivers will do start-of-VBI, e.g. the instant that a refresh cycle ends and it enters the idle period between refresh cycles (aka the VBI, the blanking interval).    So we can assume that first in abscence of information.  But we should verify.

11. Be noted that there could be an offset in the stationary tearline in step 8-9 depending on if the API is heartbeating on the beginning-of-VBI or end-of-VBI.  

12. First, we want to access what the current timings of the current mode.  This can be determined by simple tests with signal utilities for creating custom signal timings (Linux [Modeline](https://en.wikipedia.org/wiki/XFree86_Modeline), [xrandr](https://wiki.archlinux.org/title/xrandr), etc], Windows [ToastyX CRU](https://www.monitortests.com/) or [NVIDIA Custom Resolution](https://nvidia.custhelp.com/app/answers/detail/a_id/759/~/custom-resolutions), or Mac [SwitchResX](https://www.madrau.com/)

13. Create a custom ModeLine that makes visual debugging easier (1080p 60Hz mode that transmits a 60Hz refresh cycle in 1/75sec).  No equipment needed!  No oscilloscope needed!  Just your eyes.

Most ordinary 60Hz desktop computer displays (22"-27") have at least an undocumented 75Hz mode built into it.  At least all the DELL and HP office monitors do.  We will piggyback on this as a tearline-tester.   We want to create mode where you have a 60Hz mode transmitting over the video cable in roughly 1/75sec.   A normal 60Hz 1080p mode is a 1125 Vertical Total, so we need (1125 * 75 / 60) = 1406 scanlines per 1080p refresh cycle.  This will usually sync fine on most panels.   

A large VBI will make tearline-hiding debugging easier (VSYNC ON simulation via VSYNC OFF) because there's a taller VBI to hide a very jittery tearline in.  But more importantly, this data will determine if the VSYNC heartbeat behavior of a graphics driver is beginning-of-VBI or end-of-VBI behavior.

For a Linux Modeline, just test a modified 1080p 60Hz mode by editing number "1125" to "1406" for a standard 1080p 60Hz mode.  This will not sync on all displays so be warned, and be prepared to fallback / immediately exit to revert change.  This is a temporary change only, for one-time debugging.   For Windows, use NVIDIA Control Panel, you can edit "1125" to "1406" in the NVIDIA Custom Resolution

![image](https://user-images.githubusercontent.com/59981975/218231805-c9542640-2c50-407a-a7a8-7ae4fdcc6c27.png)

(Portion of a helper image used for a different purpose, but illustrates the Windows version of a video signal customization)

Once this is done, you've got 1080 visible pixel rows and 326 offscreen pixel rows, which now means your display is spending only 77% of its time refreshing (1080/1406th), and 23% of its time idling between refresh cycles (326/1406th).   It's only a minor QFT speedup of a 60Hz refresh cycled delivery, but something magical happens: Easy visual measurement of the time-offset of frame presentation.

Small VBIs of only 45 pixels can make tearline-testing/debugging more difficult especially on slower/internal GPUs, so analyzing this is easier on a large-vertical-total mode where the VBI is massively larger than the vertical amplitude of a vibrating tearline (timing error / raster jitter).

We already know the horizontal scanrate of a 1406-line 60Hz mode: You multiply the vertical total by refresh rate (1406 x 60) = 84360  ... That's almost the same horizontal scan rate of a 75 Hz mode (1125 x 75) = 84375 so most 75Hz-capable LCDs will sync to this custom 60Hz mode.

Now repeat the test in a modified version of 8-9:

15. Run software in VSYNC OFF.  Run that panning test in VSYNC OFF mode.  Do a custom precise busyloop then present rendered frame (of panning bitmap) and immediately flush at the spot.  Make sure we're running on this custom signal (84360 scanlines).  With two busywait tests:
A. Busy-wait-loop by 100 scanlines (Busyloop for 100/84360 second after VSYNC heartbeat) 
B. Busy-wait-loop VBI+100 scanlines (Busyloop for 426/84360 second after VSYNC heartbeat).  

17. Visual observation is a confirm.  Make sure tearline is jittering by less than 100 tearline amplitude.  Keep your render simple (e.g. just one bitmap with vertical 3edges, horizontally panning fast, such as 960 pixels/sec), for easy tearing-detection.  
A. If 100-scanline-delay has invisible tearline, and 426-scanline-delay has a stationary tearline approximately at pixel row #100 then you've confirmed **VSYNC heartbeat is END OF VBI**
B. If 100-scanline-delay has a tearline near pixel row #100, and 426-scanline-delay has a tearline near pixel row #426, then you've confirmed **VSYNC heartbeat is START OF VBI**

18. Compensate accordingly.

19. Now we've got the hallmarks of an API that provides easier cross platform beam racing (whether single frame based, or subrefresh-based)

**Comment by mdrejhon** [comment]
BTW, here's a great piece on Android-based beam racing (aka modern equivalent of generic raster interrupts on Android)

https://blog.imaginationtech.com/reducing-latency-in-vr-by-using-single-buffered-strip-rendering/

(*However, the concept is the same on every device that connects to a display through HDMI, DisplayPort, eDP, DVI, VGA, component, composite, -- or even the ribbon cable between a smartphone motherboard and a smartphone display.  Virtually every raster graphics 2D display are actually beam raceable, given appropriate access to minimum data required.*)

**Comment by mdrejhon** [comment]
Update -- 

I am looking into the possibility of making this portion an independent repo (e.g. a separate project that is compatible with GLFW but also other graphics frameworks).   

I have a network of open source beamracer programmer that I can probably reach out (Tom Harte does raster beamracing on Macs, and I do beamracing on PCs, and I know two emulator authors that raster beamraces on PCs) to merge the knowledge into the world first cross-platform vsync heartbeat listener / beam racing library.  

This might reduce GLFW responsibilities to just:

* Providing a true tearing-visible glfwSwapInterval(0) mode.
* Providing offscreen glfw rendering support (this is needed for some pipelines), e.g. invisible window that still works correctly

Then the being-incubated crossplatform beam racing library (probably MIT license, BSD0 license, or Apache license).  Blur Busters has gigantic need for this project that I currently offer four-figure bounty funding.

**Comment by BlackCapCoder** [comment]
Hi!
This thread was helpful for finding a Linux solution, so I thought I'd share my code in case anyone finds it useful.

Sorry for the length, but I've cut out everything that is not relevant and made it read top-to-bottom, if you can believe it..

```cpp
#include <GL/gl.h>
#define GLX_GLXEXT_PROTOTYPES
#include <GL/glx.h>


struct Renderer
{
  Renderer
    ( Display *  dpy
    , Window     win
    , Drawable   xdr
    ) : dpy{dpy}, win{win}, xdr{xdr}
  {
    initSyncVars ();
  }

  // These are initialized by `initSyncVars`:

  // The time between screen refreshes, in microseconds,
  // is given by `mscA + mscB / mscC`, where `mscB / mscC`
  // is some rational number less than 1
  int32_t mscA;
  int32_t mscB;
  int32_t mscC;

  // Before we begin drawing, we wait for a screen refresh and
  // take not of take note of ..

  int64_t ust_init; // current time in microseconds
  int64_t msc_init; // total number of screen refreshes (globally)


  // -------- Just for convenience:
  
  int64_t ust, msc, sbc;

  void updateTime ()
  {
    glXGetSyncValuesOML (dpy, win, &ust, &msc, &sbc);
  }

  int waitForMscOML (int target_msc, int divisor, int remainder)
  {
    return glXWaitForMscOML (dpy, win, target_msc, divisor, remainder, &ust, &msc, &sbc);
  }

  // ----------


  void initSyncVars ()
  {
    int32_t a0, b0;
    glXGetMscRateOML (dpy, xdr, &a0, &b0);

    // double rate_hz = (double) a0 / b0;  //   60.0204  monitor refresh rate
    // double rate_ms = 1000.0 / rate_hz;  //   16.6610  ms between refreshes
    // double rate_us = 1000.0 * rate_ms;  //  ~16661    microseconds
    //
    // rate_us = (1000000.0 * (double) b0) / a0

    int32_t f  = gcd (1000000, a0); // to keep the numbers nice and small
    int32_t a1 = a0 / f;
    int32_t b1 = b0 * (1000000 / f);

    mscA = b1 / a1; // a2
    mscB = b1 % a1; // b2
    mscC = a1;      // c

    // b0/a0*1000000 = a2 + b2/c
    // a1/b1*1000000 = a0/b0

    // ---

    updateTime ();
    waitForMscOML (0, 0, 0);
    waitForMscOML (msc+1, 0, 0);

    msc_init = msc;
    ust_init = ust;
  }

  // ----

  constexpr int64_t calcRefreshTime (int64_t msc, double err)
  {
    return
      msc * mscA + std::floor (
        msc * (err + ((double) mscB / mscC))
      );
  }

  // Note the `err` parameter.
  //
  // The `ust` (current time) value produced by waitForMscOML
  // is very close to the time of the redraw, but it is
  // either not exact, amd/or the error introduced by floor
  // adds up over time. `err` is a workaround.


  // Error correction for refresh guesses
  double guessErrF = 0;

  // Call this once you know the `ust` that the `msc`th
  // refreh occed at.
  void reportRefreshTime (int64_t ust, int64_t msc)
  {
    int64_t delta_ust = ust - ust_init;
    int64_t delta_msc = msc - msc_init;

    // What we would have guessed with no error correction
    int64_t guess =
      calcRefreshTime (delta_msc, 0);

    // +-microseconds that guess would have been off
    int64_t errI =
      delta_ust - guess;

    // the error correction value that would have 
    // made the guess correct
    double errF =
      (double) errI / delta_msc;


    // Either one of these is fine; I don't know if one is better:

    // guessErrF = errF;
    guessErrF = (guessErrF + errF)/2;

    // If `waitForMscOML` where to report a time that is "way off"
    // the former would make our next guess also way off.
    // The later is intended to alleviate this.
  }


  int64_t guessTimeOfNextRefresh ()
  {
    int64_t delta_msc = msc - msc_init;

    return ust_init +
      calcRefreshTime (delta_msc + 1, guessErrF);
  }

  // On my system, ^ this function seems to be about
  // in the range of +-10us of the actual refresh.
  // +-3us is pretty typical, but it can make the
  // occasional bad guess.

  // This is more than enough for me though, and certanly
  // better than SwapBuffer which would block for MILLISECONDS

  // ----

  // Here's my rendering loop:

  // I call this in the main thread- less sdl would misbehave-
  // and have my event processing loop (using X, which is thread safe)
  // in a separate thread.


  // all rendering needs to happen on this thread,
  // so I have a queue of things to be rendered
  TQueue <Task> work {};

  // ..and some arbitrarily chosen deadline before the
  // refresh, after which I will no longer be accepting
  // new tasks. Overshooting the refresh would be terrible,
  // so I'd rather be safe than sorry.
  int renderDeadline = 50;


  void renderLoop ()
  {
    while (!shouldQuit)
    {
      if (work.empty ())
        work.snooze (); // block until we have something to draw
      work.pop () (); // draw it- now we're on a timer

      updateTime ();
      const int64_t deadline = 
        guessTimeOfNextRefresh () - renderDeadline;

      // keep drawing stuff until the deadline
      while (ust > deadline)
      {
        if (work.empty ())
        {
          bool ok = work.snooze (deadline - ust); // block with timeout
          if (!ok) break; // out of time!
        }
        work.pop () ();
        updateTime ();
      }

      glFlush ();
      waitForMscOML (msc+1, 0, 0);
      // glXSwapBuffers (dpy, xdr);

      reportRefreshTime (ust, msc);
    }
  }
};
```

**Comment by ad8e** [comment]
There's a repo https://github.com/ad8e/vsync_blurbusters/ that solves the issue without needing to wait for the vblank. You can see the Linux part at https://github.com/ad8e/vsync_blurbusters/blob/main/platform_vsync_linux.cpp. The system gives you the timing.

I actually think this issue is no longer necessary, since the problem is solved on both Windows and Linux. glfw provided enough of the underlying system handles to make things convenient. This issue can be closed if anyone wants, or I can close it myself if there are no objections.

**Comment by mdrejhon** [comment]
Thanks!  Yes, that's true.  And there's also the alternative cross platform algorithm at https://github.com/blurbusters/RefreshRateCalculator which is very generic and just needs to listen to a noisy VSYNC heartbeat (jittered / dropped) to average-out very accurate VSYNC's and raster estimates.   

The AD8E algorithm is better, but the RefreshRateCalculator algorithm is currently deployed by www.vsynctester.com and www.testufo.com/refreshrate -- so there are multiple options now;
